types/react/index.d.ts

```diff
@@ -558,7 +632,45 @@ declare namespace React {
 
     function createRef<T>(): RefObject<T>;
 
-    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
+    // will show `ForwardRef(${Component.displayName || Component.name})` in devtools by default,
+    // but can be given its own specific name
+    interface ForwardRefExoticComponent<P> extends NamedExoticComponent<P> {
+        defaultProps?: Partial<P>;
+    }
+
+    function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ForwardRefExoticComponent<P & ClassAttributes<T>>;
+
+    type ComponentProps<T extends ComponentType<any>> =
+        T extends ComponentType<infer P> ? P : {};
+    type ComponentPropsWithRef<T extends ComponentType<any>> =
+        T extends ComponentClass<infer P>
+            ? P & ClassAttributes<InstanceType<T>>
+            : T extends SFC<infer P>
+                ? P
+                : {};
+
+    // will show `Memo(${Component.displayName || Component.name})` in devtools by default,
+    // but can be given its own specific name
+    interface MemoExoticComponent<T extends ComponentType<any>> extends NamedExoticComponent<ComponentPropsWithRef<T>> {
+        readonly type: T;
+    }
+
+    function memo<P extends object>(
+        Component: SFC<P>,
```

Comment:
Is this right? If I have a render prop component (where children might be a function), it's not being inferred as functional component.

Reply:
Implicit `children` strike back

Again

🙄

I guess I have a New Years break project for a breaking change. Hopefully one I can get to compile.

Reply:
I'm investigating it further. The problem may not be `children`, but something else (my current bet is functional components with custom generic types). I'll see if I can come up with a minimum reproduction.

Reply:
@sandersn @johnnyreilly Would it be possible to condition this breaking change on a specific TypeScript version through typesVersion?

I was thinking of being able to make the change incremental by only updating types on demand, when people update to, say, TS3.3. This would require kind of a "max supported TS version" on types that depend on `@types/react`.

I'm a bit afraid of it being a python3 problem, though, but the technical debt of `@types/react` is now nigh-unmanageable.

Yes, I could work around this problem in this specific instance by changing the input to a function instead of FunctionComponent, but it's pervarsive and everywhere. A lot of things that currently work in the ecosystem (like function children) work because of bugs in the types, not because of features.

Reply:
@diegohaz ah, yes, that is the one big problem with mapped types (and the lack of kinds?); they erase generics and collapse unions. If the props are either of those the output type may be funky.

Reply:
In another thread we decided it was time to update the types to expect TS 3

Reply:
> Would it be possible to condition this breaking change on a specific TypeScript version through typesVersion?

I'm afraid I haven't played with `typesVersion` yet and so can't advise. That said, as @ericanderson mentioned there's been prior discussion about upgrading to TS 3. We should do that anyway. If anyone is thinking about submitting a PR that does that then they should expect ♥️ and 🤗